src/Shared/InterningBinaryReader.cs

```diff
@@ -152,57 +166,97 @@ override public String ReadString()
         /// <summary>
         /// A shared buffer to avoid extra allocations in InterningBinaryReader.
         /// </summary>
+        /// <remarks>
+        /// The caller is responsible for managing the lifetime of the returned buffer and for passing it to <see cref="Create"/>.
+        /// </remarks>
         internal static SharedReadBuffer CreateSharedBuffer()
         {
             return new Buffer();
         }
 
+        /// <summary>
+        /// Gets a buffer from the pool or creates a new one.
+        /// </summary>
+        /// <returns>The <see cref="Buffer"/>. Should be returned to the pool after we're done with it.</returns>
+        private static Buffer GetPooledBuffer()
+        {
+            Buffer buffer = Interlocked.Exchange(ref s_bufferPool, null);
+            if (buffer != null)
+            {
+                return buffer;
+            }
+            return new Buffer();
```

Comment:
**NIT**
This line is called in following cases:
1) very 1st pooled buffer creation
2) 'not using using' bug in code - no dispose called
3) using BinaryReadTranslator concurrently

Both 2. and 3. invalidate usability of pooled buffer. If using BinaryReadTranslator concurrently is rare or not expected at all, I'd like to have some debug only heuristics warnings about such cases. Something in line of:
```C#
[Conditional("DEBUG")]
public void EnsureNotCalledVeryOften(TimeSpan range, int maxCount)
{
  lock(s_me)
  {
    if (DateTime.UtcNow - s_lastMeasuredRangeStart > range)
    {
       s_lastMeasuredRangeStart = DateTime.UtcNow;
       s_occurenceCount = 0;
    }
    if (++s_occurenceCount > maxCount)
     throw new InvalidOperationException("Suspiciously often created PooledBuffer. Please verify proper Disposing.")
  }
}
```

Reply:
This is a good point. Especially given our aspiration to remove the single-threadedness at some point. Though I'm not sure if such a complex invariant is worth capturing in the code or if we should rely on testing infrastructure to catch regressions reliably and early. For 2. that would be something like a static code analyzer, for 3. a performance benchmark (catching all regressions, not just ones we foresee - in this particular case it is covered with RPS so I believe it would catch it).